up_client_socket/rust/u_transport_socket.rs

```diff
@@ -0,0 +1,395 @@
+use up_rust::{UCode, UMessage, UMessageType, UStatus, UTransport, UUri};
+
+use protobuf::Message;
+use std::net::TcpStream as TcpStreamSync;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use std::io::{Read, Write};
+
+use crate::constants::BYTES_MSG_LENGTH;
+use crate::constants::DISPATCHER_ADDR;
+
+pub trait UtransportExt {
+     fn socket_init(&mut self);
+     fn _handle_publish_message(&mut self, umsg: UMessage);
+     fn _handle_request_message(&mut self, umsg: UMessage);
+    fn read_socket(&self, buffer: &mut [u8]) -> io::Result<usize>;
+}
+
+pub struct UtransportSocket {
+    socket_sync: TcpStreamSync,
+    listner_map: Arc<Mutex<HashMap<String, Vec<Arc<dyn UListener>>>>>,
+}
+impl Clone for UtransportSocket {
+    fn clone(&self) -> Self {
+        UtransportSocket {
+            socket_sync: self
+                .socket_sync
+                .try_clone()
+                .expect("issue in cloning sync socket"),
+            listner_map: self.listner_map.clone(),
+        }
+    }
+}
+
+impl UtransportSocket {
+    pub fn new() -> Self {
+        let socket_sync: TcpStreamSync =
+            TcpStreamSync::connect(DISPATCHER_ADDR).expect("issue in connecting  sync socket");
+
+        UtransportSocket {
+            socket_sync,
+            listner_map: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+impl UtransportExt for UtransportSocket {
+     fn socket_init(&mut self) {
+        loop {
+            // Receive data from the socket
+            let mut buffer: [u8; BYTES_MSG_LENGTH] = [0; BYTES_MSG_LENGTH];
+
+            let bytes_read = match self.read_socket(&mut buffer) {
+                Ok(bytes_read) => bytes_read,
+                Err(e) => {
+                    dbg!("Socket error: {}", e);
+                    break;
+                }
+            };
+
+            // Check if no data is received
+            if bytes_read == 0 {
+                continue;
+            }
+
+            let umessage = UMessage::parse_from_bytes(&buffer[..bytes_read]).expect("Failed to parse message");
+
+            match umessage.attributes.type_.enum_value() {
+                Ok(mt) => match mt {
+                    UMessageType::UMESSAGE_TYPE_PUBLISH => {
+                        println!("calling handle publish...");
+                        self._handle_publish_message(umessage);
+                        ()
+                    }
+                    UMessageType::UMESSAGE_TYPE_NOTIFICATION => todo!(),
+                    UMessageType::UMESSAGE_TYPE_UNSPECIFIED => (),
+                    UMessageType::UMESSAGE_TYPE_RESPONSE => (),
+                    UMessageType::UMESSAGE_TYPE_REQUEST => {
+                        self._handle_request_message(umessage);
+                        ()
+                    }
+                },
+                Err(_) => (),
+            }
+        }
+    }
+
+    fn read_socket(&self, buffer: &mut [u8]) -> io::Result<usize> {
```

Comment:
If you follow [this suggestion](https://github.com/eclipse-uprotocol/up-tck/pull/23/files#r1566307476), then this method goes away.